File: iosApp/iosApp/Modules/CurrencySwap/CurrencySwapViewModel.swift
// Created by web3d4v on 14/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation

struct CurrencySwapViewModel {
    let title: String
    let items: [Item]
}

extension CurrencySwapViewModel {
    
    enum Item {
        case swap(Swap)
        case limit
    }
    
    struct Swap {
        
        let currencyFrom: CurrencyEnterAmountViewModel
        let currencyTo: CurrencyEnterAmountViewModel
        let currencySwapProviderViewModel: CurrencySwapProviderViewModel
        let currencySwapPriceViewModel: CurrencySwapPriceViewModel
        let currencySwapSlippageViewModel: CurrencySwapSlippageViewModel
        let currencyNetworkFeeViewModel: NetworkFeePickerViewModel
        let isCalculating: Bool
        let providerAsset: String
        let approveState: ApproveState
        let buttonState: ButtonState

        enum ApproveState {
            case approve
            case approving
            case approved
        }
        
        enum ButtonState {
            case loading
            case invalid(text: String)
            case swap
            case swapAnyway(text: String)
        }
    }
}

extension Array where Element == CurrencySwapViewModel.Item {
    var swap: CurrencySwapViewModel.Swap? {
        var swap: CurrencySwapViewModel.Swap?
        forEach { if case let CurrencySwapViewModel.Item.swap(value) = $0 { swap = value } }
        return swap
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencySwap/CurrencySwapViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.currencySwap

import com.sonsofcrypto.web3lib.formatters.Formatters
import com.sonsofcrypto.web3walletcore.common.viewModels.*

data class CurrencySwapViewModel(
    val title: String,
    val items: List<Item>,
) {
    sealed class Item {
        data class Swap(val swap: SwapData): Item()
    }

    data class SwapData(
        val currencyFrom: CurrencyAmountPickerViewModel,
        val currencyTo: CurrencyAmountPickerViewModel,
        val currencySwapProviderViewModel: CurrencySwapProviderViewModel,
        val currencySwapPriceViewModel: CurrencySwapPriceViewModel,
        val currencySwapSlippageViewModel: CurrencySwapSlippageViewModel,
        val currencyNetworkFeeViewModel: NetworkFeeViewModel,
        val isCalculating: Boolean,
        val providerAsset: String,
        val approveState: ApproveState,
        val buttonState: ButtonState,
    )

    enum class ApproveState { APPROVE, APPROVING, APPROVED }

    sealed class ButtonState {
        object Loading: ButtonState()
        data class Invalid(val text: String): ButtonState()
        object Swap: ButtonState()
        data class SwapAnyway(val text: String): ButtonState()
    }
}


data class CurrencySwapPriceViewModel(
    val value: List<Formatters.Output>
)

data class CurrencySwapProviderViewModel(
    val iconName: String,
    val name: String,
)

data class CurrencySwapSlippageViewModel(
    val value: String
)
